### DISCUSSION TITLE
Automatic grouping for saved links: measure before building

### DISCUSSION THREAD

**Discussion by mvoronin** [mvoronin/trellmark#46] [Ideas]
Should saved links be filed into groups automatically, for hosts the existing domain rules do not cover?

Opening this as a discussion rather than an issue on purpose. The honest answer may be that this feature has **no job at all**, and four SQL queries decide which. An issue asserts work will happen and carries a closing condition; this does not have one yet.

## Automatic filing already ships

This is worth stating first, because it narrows the question a great deal. `PostgresBookmarkRepository.create_url` (`trellmark/bookmarks/persistence.py:429`) already files every saved link:

1. Resolve the hostname with `domain_for_url` (`trellmark/bookmarks/domain.py:398`), which returns `None` rather than raising for hosts that are not matchable domains.
2. Add the URL to **every** group holding a matching `group_domains` row.
3. If nothing matched, fall back to the group literally named `default` (`persistence.py:445`), which is protected from rename and deletion.

So there is nothing to build at the "route by domain" tier. It exists, it is exact-hostname, it is user-authored, and it is deterministic. The real gap is narrower than "sort my links": it is **hosts with no rule, and hosts whose rule is ambiguous**.

That gap has an observable name. A link landing in `default` is the system saying *I had no idea*. **The size of `default` is the size of the problem.**

## An LLM is not the tool here

This is classification into an existing taxonomy — not clustering, not generation. The taxonomy is personal, lives entirely in this database, and no general-purpose model has ever seen it. A counting model trained on the user's own filing history beats an LLM on latency, determinism, explainability, and — because it learns *this* taxonomy rather than guessing at a plausible one — very likely on accuracy too. It also keeps the "no external service" property that everything else in this project has.

## Two schema facts that shape any answer

| Fact | Consequence |
|---|---|
| A URL may belong to **many** groups — `url_groups` has composite PK `(url_id, group_id)` and `urls` has no `group_id` | This is **multi-label**, not multi-class. The output is a ranked set with a threshold, never an `argmax` |
| `urls.title` is **nullable** | If most rows have no title, a text model sees only URL tokens — roughly the information the hostname already carries, which collapses the text tier back into the counting tier |

## The ground truth is already in the database

There is no audit trail of moves, and none needs to be added. For any saved URL, recompute what the domain rules *would* assign and diff that against actual membership in `url_groups`. Where they differ, the user overrode the rule.

That difference is the **only** signal carrying information the deterministic tier does not already have. Training on rule-derived memberships teaches a model to reproduce a lookup table that is already exact — impressive-looking accuracy, zero added value.

## The tiers, if the numbers justify any of them

| Tier | How | Cost | Where it breaks |
|---|---|---|---|
| **0. Learned host statistics** | Frequency table `host → group` over saved links. A host seen at least *N* times with a strong majority in one group becomes a suggestion — effectively proposing the `group_domains` rule the user never wrote | none | Hosts that legitimately feed several groups: YouTube, Reddit, Medium, GitHub. Measure this before assuming it is rare |
| **1. Linear text model** | Multinomial naive Bayes or logistic regression over namespaced features — `host:`, `rdom:`, `tld:`, `path:`, `title:` — with add-α smoothing, one-vs-rest per group for the multi-label shape. Training over a few thousand links takes milliseconds | ~150 lines of plain Python, or scikit-learn if the numpy/scipy weight is acceptable | Short or missing titles; heavy class imbalance against a large catch-all group. Naive Bayes scores are badly calibrated, so threshold on the **top-1 to top-2 margin**, not on a raw probability |
| **2. Embeddings + nearest centroid** | One vector per link from title plus description; suggest the group whose centroid is nearest. Brute-force cosine in numpy is fine well past 100k links — **no pgvector needed** | a model, 100–500 MB | The runtime, not the maths. Keeping it out of the app image means a sidecar in the existing pod speaking HTTP over localhost — still no external service, and the feature degrades to tier 1 when the sidecar is absent |

**Clustering is a different feature.** k-means or HDBSCAN over embeddings does not route one link; it answers "what new groups should exist for this unsorted pile". Useful occasionally, on the `default` group, as an explicit action — not part of this.

## Behaviour this has to commit to

- **The deterministic layer always wins.** Precedence: explicit `group_domains` rule → learned host statistics → text model → abstain.
- **Never auto-move.** The link stays in `default` and a suggestion appears beside it. Accepting one is an ordinary move, and the next rebuild picks it up as a labeled example. No feedback table, no online learning.
- **Abstention is correct behaviour, not failure.** A missing suggestion costs a manual move the user was making anyway; a wrong suggestion that gets accepted quietly misfiles data.
- **Return the reason.** Feature contributions are nearly free to extract from naive Bayes, and turn a bad suggestion into an actionable moment — *suggested because of `rdom:crates.io`, `path:docs`* — which is exactly the prompt to write a `group_domains` rule instead and remove the guesswork.
- **Respect Safe mode.** It is enforced in the browser and the server has no NSFW filter anywhere, so a response must carry each candidate's `nsfw` flag and let the client filter. Otherwise a suggestion surfaces a group the user deliberately hid.

## The measurement — needs no plan and no code

Four queries decide whether this feature has a job. Post results as replies.

**1. How big is the problem?**

```sql
SELECT count(*)
FROM url_groups ug
JOIN groups g ON g.id = ug.group_id
WHERE lower(g.name) = 'default';
```

**2. Is there text to learn from?**

```sql
SELECT count(*) FILTER (WHERE title IS NOT NULL) AS titled,
       count(*)                                  AS total
FROM urls;
```

**3. How much of the filing was a human decision rather than a rule?**

The hostname extraction here is a rough stand-in for `domain_for_url()` — no IDNA, no IP normalization — so treat this as an estimate.

```sql
WITH hosts AS (
    SELECT id AS url_id,
           lower(substring(url FROM '^https?://([^/:?#]+)')) AS host
    FROM urls
),
predicted AS (
    SELECT h.url_id, gd.group_id
    FROM hosts h
    JOIN group_domains gd ON gd.domain = h.host
),
expected AS (
    SELECT url_id, group_id FROM predicted
    UNION ALL
    SELECT h.url_id, (SELECT id FROM groups WHERE lower(name) = 'default')
    FROM hosts h
    WHERE NOT EXISTS (SELECT 1 FROM predicted p WHERE p.url_id = h.url_id)
)
SELECT count(DISTINCT url_id) AS overridden
FROM (
    (SELECT url_id, group_id FROM url_groups
     EXCEPT SELECT url_id, group_id FROM expected)
    UNION
    (SELECT url_id, group_id FROM expected
     EXCEPT SELECT url_id, group_id FROM url_groups)
) AS diff;
```

**4. Is there enough data per label?**

```sql
SELECT g.name, count(ug.url_id) AS links
FROM groups g
LEFT JOIN url_groups ug ON ug.group_id = g.id
GROUP BY g.id, g.name
ORDER BY links DESC;
```

Worth adding once those are in: how many hostnames appear under more than one group. That number alone says whether tier 0 is sufficient, or whether tier 1 is carrying the feature.

## Decision rule

**If `default` is near-empty, or the override count in query 3 is small, there is nothing to learn and this feature has no job yet.** Building it anyway produces a subsystem whose measurable contribution over the existing domain rules is noise.

That outcome closes this discussion honestly — it is not a failure of the idea. It is also worth saying plainly that this is new product work outside the current Notes MVP scope in `.planning/PROJECT.md`, so even a favourable measurement schedules nothing by itself.
